src/crypto.ts

```diff
@@ -195,4 +200,57 @@ export default class Crypto {
       decodeBase64(formSecretKey)
     )
   }
+
+  /**
+   * Decrypts an encrypted submission, and also download and decrypt any attachments alongside it.
+   * @param formSecretKey Secret key as a base-64 string
+   * @param decryptParams The params containing encrypted content and information.
+   * @returns A promise of the decrypted submission, including attachments (if any). Or else returns null if a decryption error decrypting any part of the submission.
+   * @throws {MissingPublicKeyError} if a public key is not provided when instantiating this class and is needed for verifying signed content.
+   */
+  decryptWithAttachments = async (
+    formSecretKey: string,
+    decryptParams: DecryptParams
+  ): Promise<DecryptedContentAndAttachments | null> => {
+    const decryptedRecords: DecryptedAttachments = {}
+    const filenames: Record<string, string> = {}
+
+    const attachmentRecords: EncryptedAttachmentRecords = decryptParams.attachmentDownloadUrls ?? {}
+    const decryptedContent = this.decrypt(formSecretKey, decryptParams)
+    if (decryptedContent === null) return null
+
+    // Retrieve all original filenames for attachments for easy lookup
+    for (const i in decryptedContent.responses) {
+      const response = decryptedContent.responses[i]
+      if (response.fieldType === 'attachment' && response.answer) {
+        filenames[response._id] = response.answer
+      }
+    }
+
+    for (let fieldId in attachmentRecords) {
+      const downloadResponse = await axios.get(attachmentRecords[fieldId], { responseType: 'json' })
+      if (downloadResponse.status !== 200) return null
+
+      const encryptedAttachment: EncryptedAttachmentContent = downloadResponse.data
+      const encryptedFile: EncryptedFileContent = {
+        submissionPublicKey: encryptedAttachment.encryptedFile.submissionPublicKey,
+        nonce: encryptedAttachment.encryptedFile.nonce,
+        binary: decodeBase64(encryptedAttachment.encryptedFile.binary),
+      }
+      const decryptedFile = await this.decryptFile(formSecretKey, encryptedFile)
+
+      // Decryption for a file failed
+      if (decryptedFile === null) return null
+
+      // Original name for the file is not found
+      if (filenames[fieldId] === undefined) return null
```

Comment:
perhaps we could clarify the intent of this block by placing it outside and using [lodash's `some` function](https://lodash.com/docs/4.17.15#some)

Reply:
agree, we can fail early outside the for-loop instead of calling `axios.get` and `decrypt` before finding out that we don't know the filename

Reply:
So I am trying not to introduce any additional dependencies into this except for `axios` since this is supposed to be a thin library, but I've moved the check out of the `Promise`s so it will fail immediately. Some promises might be created, but if we return from the function without awaiting any promises, then the promises will go out of scope and stop processing right?

Reply:
yup!